### DISCUSSION TITLE
[firebase_messaging] FirebaseMessaging.onBackgroundMessage causes the application crash

### DISCUSSION THREAD

**Discussion by abdullahrs** [firebase/flutterfire#5721] [Q&A] [answered]
Before Flutter 2 and null-security, my firebase services were working fine. I wanted to update it according to the documentation. I followed [documentation ](https://firebase.flutter.dev/docs/messaging/usage/)instructions but I get this error and the app is closing immediately. I tried many things but did not get any results. 
I just added these, am I doing something wrong? 
main.dart file:
```
// imports ...

final FlutterLocalNotificationsPlugin flutterLocalNotificationsPlugin =
    FlutterLocalNotificationsPlugin();

const AndroidNotificationChannel channel = AndroidNotificationChannel(
  'high_importance_channel', // id
  'High Importance Notifications', // title
  'This channel is used for important notifications.', // description
  importance: Importance.high,
);

Future<void> _firebaseMessagingBackgroundHandler(RemoteMessage message) async {
  await Firebase.initializeApp();
  print('Handling a background message ${message.messageId}');
}

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await UserPreferences().init();
  
  Initialize.instance.init();

  await Firebase.initializeApp();

  // ? Error occurs here
  // FirebaseMessaging.onBackgroundMessage(_firebaseMessagingBackgroundHandler);

  await flutterLocalNotificationsPlugin
      .resolvePlatformSpecificImplementation<
          AndroidFlutterLocalNotificationsPlugin>()
      ?.createNotificationChannel(channel);
      
  await FirebaseMessaging.instance.setForegroundNotificationPresentationOptions(
    alert: true,
    badge: true,
    sound: true,
  );
  //...
}
```

Error:
> 
> Launching lib\main.dart on sdk gphone x86 arm in debug mode...
> lib\main.dart:1
> √ Built build\app\outputs\flutter-apk\app-debug.apk.
> I/Choreographer( 7761): Skipped 131 frames!  The application may be doing too much work on its main thread.
> Connecting to VM Service at ws://127.0.0.1:52375/z4dqvdo3Op4=/ws
> I/FLTFireBGExecutor( 7761): Creating background FlutterEngine instance, with args: [--start-paused, --enable-dart-profiling]
> D/EGL_emulation( 7761): eglCreateContext: 0xe9d52140: maj 3 min 0 rcv 3
> D/EGL_emulation( 7761): eglCreateContext: 0xe9d50a80: maj 3 min 0 rcv 3
> D/HostConnection( 7761): HostConnection::get() New Host Connection established 0xe9d52220, tid 7827
> D/HostConnection( 7761): HostComposition ext ANDROID_EMU_CHECKSUM_HELPER_v1 ANDROID_EMU_native_sync_v2 ANDROID_EMU_native_sync_v3 ANDROID_EMU_native_sync_v4 ANDROID_EMU_dma_v1 ANDROID_EMU_direct_mem ANDROID_EMU_host_composition_v1 ANDROID_EMU_host_composition_v2 ANDROID_EMU_vulkan ANDROID_EMU_deferred_vulkan_commands ANDROID_EMU_vulkan_null_optional_strings ANDROID_EMU_vulkan_create_resources_with_requirements ANDROID_EMU_YUV_Cache ANDROID_EMU_async_unmap_buffer ANDROID_EMU_vulkan_ignored_handles ANDROID_EMU_vulkan_free_memory_sync ANDROID_EMU_vulkan_shader_float16_int8 ANDROID_EMU_vulkan_async_queue_submit GL_OES_EGL_image_external_essl3 GL_OES_vertex_array_object GL_KHR_texture_compression_astc_ldr ANDROID_EMU_host_side_tracing ANDROID_EMU_async_frame_commands ANDROID_EMU_gles_max_version_3_0
> D/EGL_emulation( 7761): eglMakeCurrent: 0xe9d50a80: ver 3 0 (tinfo 0xab9518d0) (first time)
> D/AndroidRuntime( 7761): Shutting down VM
> E/AndroidRuntime( 7761): FATAL EXCEPTION: main
> E/AndroidRuntime( 7761): Process: com.abta.mrc, PID: 7761
> E/AndroidRuntime( 7761): java.lang.ClassCastException: io.flutter.embedding.engine.plugins.shim.ShimPluginRegistry cannot be cast to io.flutter.embedding.engine.FlutterEngine
> E/AndroidRuntime( 7761): 	at com.abta.mrc.Application.registerWith(Application.java:25)
> E/AndroidRuntime( 7761): 	at io.flutter.plugins.firebase.messaging.FlutterFirebaseMessagingBackgroundExecutor.lambda$null$0$FlutterFirebaseMessagingBackgroundExecutor(FlutterFirebaseMessagingBackgroundExecutor.java:206)
> E/AndroidRuntime( 7761): 	at io.flutter.plugins.firebase.messaging.-$$Lambda$FlutterFirebaseMessagingBackgroundExecutor$Tw1_-iY8wcRJPWf1qeylRYPPEZI.run(Unknown Source:6)
> E/AndroidRuntime( 7761): 	at android.os.Handler.handleCallback(Handler.java:938)
> E/AndroidRuntime( 7761): 	at android.os.Handler.dispatchMessage(Handler.java:99)
> E/AndroidRuntime( 7761): 	at android.os.Looper.loop(Looper.java:223)
> E/AndroidRuntime( 7761): 	at android.app.ActivityThread.main(ActivityThread.java:7656)
> E/AndroidRuntime( 7761): 	at java.lang.reflect.Method.invoke(Native Method)
> E/AndroidRuntime( 7761): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
> E/AndroidRuntime( 7761): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
> I/Process ( 7761): Sending signal. PID: 7761 SIG: 9
> Lost connection to device.
> Exited (sigterm)

flutter doctor -v
>  [√] Flutter (Channel stable, 2.0.4, on Microsoft Windows [Version 10.0.19042.867], locale tr-TR)
>     • Flutter version 2.0.4 at C:\src\flutter
>     • Framework revision b1395592de (9 days ago), 2021-04-01 14:25:01 -0700
>     • Engine revision 2dce47073a
>     • Dart version 2.12.2
> 
> [√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
>     • Android SDK at C:\Users\Abdul\AppData\Local\Android\sdk
>     • Platform android-30, build-tools 30.0.3
>     • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
>     • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
>     • All Android licenses accepted.
> 
> [√] Chrome - develop for the web
>     • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe
> 
> [√] Android Studio (version 4.1.0)
>     • Android Studio at C:\Program Files\Android\Android Studio
>     • Flutter plugin can be installed from:
>        https://plugins.jetbrains.com/plugin/9212-flutter
>     • Dart plugin can be installed from:
>        https://plugins.jetbrains.com/plugin/6351-dart
>     • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
> 
> [√] VS Code (version 1.55.1)
>     • VS Code at C:\Users\Abdul\AppData\Local\Programs\Microsoft VS Code
>     • Flutter extension version 3.21.0
> 
> [√] Connected device (3 available)
>     • sdk gphone x86 arm (mobile) • emulator-5554 • android-x86    • Android 11 (API 30) (emulator)
>     • Chrome (web)                • chrome        • web-javascript • Google Chrome 89.0.4389.114
>     • Edge (web)                  • edge          • web-javascript • Microsoft Edge 88.0.705.74
> 
> • No issues found!

**Comment by abdullahrs** [comment, marked as answer]
I only deleted "onBackgroundMessage"  method. It looks like it is working without it.

**Reply by StatTark** [reply]
well done!
